### PR TITLE
Invite member

### PR DESCRIPTION
**Pull request by maltee1**
Tested with Signal only user and matrix user logged into the bridge

### CODE REVIEW

File: mautrix_signal/portal.py [hunk 1]
```diff
@@ -722,6 +722,17 @@ async def handle_matrix_leave(self, user: u.User) -> None:
             self.log.debug(f"{user.mxid} left portal to {self.chat_id}")
             # TODO cleanup if empty
 
+    async def handle_matrix_invite(self, invited_by: u.User, user: u.User) -> None:
```
Comment (tulir) [inline mautrix_signal/portal.py:725]:
```suggestion
    async def handle_matrix_invite(self, invited_by: u.User, user: u.User | p.Puppet) -> None:
```

File: mautrix_signal/matrix.py [hunk 2]
```diff
@@ -54,6 +54,16 @@ def __init__(self, bridge: "SignalBridge") -> None:
...
 
+    async def handle_invite(
+        self, room_id: RoomID, user_id: UserID, inviter: u.User, event_id: EventID
+    ) -> None:
+        user = await u.User.get_by_mxid(user_id, create=False)
+        if not user:
+            return
+        portal = await po.Portal.get_by_mxid(room_id)
+        if portal:
+            await portal.handle_matrix_invite(inviter, user)
```
Comment (tulir) [inline mautrix_signal/matrix.py:65]: This should probably catch `RejectMatrixInvite` and send an error message (or at least log and ignore rather than letting it propagate). Something like this:

```suggestion
        if portal and not portal.is_direct:
            try:
                await portal.handle_matrix_invite(inviter, user)
            except RejectMatrixInvite as e:
                await portal.main_intent.send_notice(portal.mxid, "Failed to invite {user.mxid} on Signal: {e}")
```

(needs an import for `RejectMatrixInvite`)

File: mautrix_signal/portal.py [hunk 3]
```diff
@@ -724,14 +724,18 @@ async def handle_matrix_leave(self, user: u.User) -> None:
...
         else:
             try:
                 await self.signal.update_group(
                     invited_by.username, self.chat_id, add_members=[user.address]
                 )
             except RPCError as e:
-                raise RejectMatrixInvite(e.message) from e
+                await self.main_intent.send_notice(
+                    self.mxid, f"Failed to invite {user.mxid} on Signal: {e}"
+                )
```
Comment (tulir) [inline mautrix_signal/portal.py:738]: `RejectMatrixInvite` is the correct error for ghost invites, it's just the non-bridge-ghost `handle_invite` which wasn't catching them

File: mautrix_signal/matrix.py [hunk 4]
```diff
@@ -42,7 +42,7 @@
     from .__main__ import SignalBridge
 
 
-class MatrixHandler(BaseMatrixHandler):
+class MatrixHandlerin(BaseMatrixHandler):
```
Comment (tulir) [inline mautrix_signal/matrix.py:45]: :thinking:

File: mautrix_signal/matrix.py [hunk 5]
```diff
@@ -67,11 +82,9 @@ async def handle_leave(self, room_id: RoomID, user_id: UserID, event_id: EventID
         portal = await po.Portal.get_by_mxid(room_id)
         if not portal:
             return
-
```
Comment (sumnerevans) [inline mautrix_signal/matrix.py:70]: please revert these unnecessary whitespace changes

File: mautrix_signal/matrix.py [hunk 6]
```diff
@@ -54,6 +54,21 @@ def __init__(self, bridge: "SignalBridge") -> None:
 
         super().__init__(bridge=bridge)
 
+    async def handle_invite(
```
Comment (sumnerevans) [inline mautrix_signal/matrix.py:57]: I don't think that you need this since `handle_puppet_invite` in https://github.com/mautrix/python/blob/master/mautrix/bridge/matrix.py#L306 should handle invites for you and call the portal function automatically, but I'm not 100% sure if that's how it works.

File: mautrix_signal/portal.py [hunk 7]
```diff
@@ -725,6 +725,17 @@ async def handle_matrix_leave(self, user: u.User) -> None:
...
+    async def handle_matrix_invite(self, invited_by: u.User, user: u.User | p.Puppet) -> None:
+        if self.is_direct:
+            raise RejectMatrixInvite("You can't invite additional users to private chats.")
+        else:
+            try:
+                await self.signal.update_group(
+                    invited_by.username, self.chat_id, add_members=[user.address]
+                )
+            except RPCError as e:
+                raise RejectMatrixInvite(e.message) from e
```
Comment (sumnerevans) [inline mautrix_signal/portal.py:737]:
```suggestion

        try:
            await self.signal.update_group(
                invited_by.username, self.chat_id, add_members=[user.address]
            )
        except RPCError as e:
            raise RejectMatrixInvite(e.message) from e
```

**Review by sumnerevans — DISMISSED** [review]

File: mautrix_signal/matrix.py (hunk 6 above)
Comment (tulir) [reply mautrix_signal/matrix.py:57]: This is for handling invites to users (rather than ghosts), so it is necessary currently

**Review by tulir — APPROVED** [review]

File: mautrix_signal/matrix.py (hunk 5 above)
Comment (maltee1) [reply mautrix_signal/matrix.py]: done
